src/sparse_blas/backends/mkl_common/mkl_helper.hpp

```diff
@@ -1,56 +1,102 @@
-* Copyright 2023 Codeplay Software Ltd.
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
 *
-* Licensed under the Apache License, Version 2.0 (the "License");
-* you may not use this file except in compliance with the License.
-* You may obtain a copy of the License at
+*      http://www.apache.org/licenses/LICENSE-2.0
 *
-* http://www.apache.org/licenses/LICENSE-2.0
+*  For your convenience, a copy of the License has been included in this
+*  repository.
 *
-* Unless required by applicable law or agreed to in writing,
-* software distributed under the License is distributed on an "AS IS" BASIS,
-* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-* See the License for the specific language governing permissions
-* and limitations under the License.
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
 *
-*
-* SPDX-License-Identifier: Apache-2.0
-*******************************************************************************/
+**************************************************************************/
 
-// MKLCPU and MKLGPU backends include
-// This include defines its own oneapi::mkl::sparse namespace with some of the types that are used here: matrix_handle_t, index_base, transpose, uolo, diag.
-#include <oneapi/mkl/spblas.hpp>
+#ifndef _ONEMKL_SRC_SPARSE_BLAS_BACKENDS_MKL_COMMON_MKL_HELPER_HPP_
+#define _ONEMKL_SRC_SPARSE_BLAS_BACKENDS_MKL_COMMON_MKL_HELPER_HPP_
 
-// Includes are set up so that oneapi::mkl::sparse namespace refers to the MKLCPU and MKLGPU backends namespace (oneMKL product)
-// in this file.
-// oneapi::mkl::sparse::detail namespace refers to the oneMKL interface namespace.
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
 
+#include "oneapi/mkl/exceptions.hpp"
 #include "oneapi/mkl/sparse_blas/detail/helper_types.hpp"
 
+#include "sparse_blas/enum_data_types.hpp"
+#include "sparse_blas/macros.hpp"
+
 namespace oneapi::mkl::sparse::detail {
 
-inline auto get_handle(detail::matrix_handle **handle) {
-    return reinterpret_cast<oneapi::mkl::sparse::matrix_handle_t *>(handle);
+/// Return whether a pointer is accessible on the host
+template <typename T>
+inline bool is_ptr_accessible_on_host(sycl::queue &queue, const T *host_or_device_ptr) {
+    auto alloc_type = sycl::get_pointer_type(host_or_device_ptr, queue.get_context());
+    // Note sycl::usm::alloc::host may not be accessible on the host according to SYCL specification.
```

Comment:
Wait, what? How can `sycl::usm::alloc::host` _not_ be host-accessible? Could you please point us to the exact section number where this is documented in the SYCL specification?

Reply:
I am referring to the table 96 here https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_kinds_of_unified_shared_memory
It says the host allocation is in host memory but is accessible by "a device" (which in this context is not necessarily host) whereas the shared allocation is clearly required to be accessible on host. I have discussed this with other SYCL developers and specification experts and we have agreed this is an oddity in the SYCL spec. I also don't recall whether we found practical applications of such an allocation. I don't think this an issue in practice. People only use shared or device allocation AFAIK.

Reply:
Well, [Table 97](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.USM.allocation.characteristics) just a little bit below that explicitly mentions that allocation type `host` is accessible on the `host`.

Reply:
That's a good point, thanks! Fixed in https://github.com/oneapi-src/oneMKL/pull/500/commits/c2b89f5580fb059a8070fb88cbe70574b59da799